### ISSUE TITLE
Questions about how lambdas are obtained

### ISSUE THREAD

**Issue by haruishi43** [cvlab-stonybrook/Beyond-Pixels#8]
Thank you for open-sourcing the code. I have some questions regarding the paper and the code:
1. At the bottom of page 7, you wrote $\gamma = \text{Softmax}(q^r_u)$. I don't understand why Softmax is used instead of Sigmoid to obtain the weights, even though the task is multi-label classification. In the code, the closest part I could find was where you computed `mask_patches = F.softmax(mask_patches, dim=2)` and this is not related to $\gamma$ because this computes labeled multi-label loss, which also raises the question of why Softmax is being used here.
2. Are there equations in the paper which corresponds to what `compute_batch_loss` is doing? It seems to me that this function is mainly used to return patches/classes where there are the most uncertainty by comparing loss values, but I couldn't completely understand what is going on inside.
3. I'm guessing the returned `mlabel_correction_idx_u` from `compute_batch_loss` are used to mask uncertain regions for multi-label classification task which I think corresponds to $\lambda_m$. This is also being used to mask out pseudo-label segmentation masks with `200`. What is this code doing?
```
class_u_w_fp[:, u_patch_x:v_patch_x, u_patch_y:v_patch_y][
    mask_u_w[:, u_patch_x:v_patch_x, u_patch_y:v_patch_y]
    == cons_class
] = 200
```
What is `cons_class` here and why is it a single class? In a multi-label setting, could there be multiple classes that are uncertain? Is this section where $\lambda_s$ is calculated?
4. What are `LL-Cp`, `LL-R`, and `LL-Ct`?
5. What are `delta_rel` and `all_class_delta_rel`? And why is `all_class_delta_rel` being updated every epoch?

**Comment by prantikhowlader** [comment]
Thanks for reaching out, I am having a few upcoming deadlines. I will provide answers to your questions  the first thing after that

**Comment by haruishi43** [comment]
@prantikhowlader thanks for the reply.
I'm gone through the code in more detail and also trained with your code, but I could not reproduce your results (achieved only 75.8 with ResNet50 for Cityscapes 1/16 split).
I've also noticed that for `all_class_delta_rel` is not being updated and is kept at consistent values for all classes of `0.000001` because `total_all_class_labeled_unconf` is not updated for each iteration. Is this intentional?

**Comment by prantikhowlader** [comment]
all_class_delta_rel[_class_]+=(((max_val-last_val)/max_val)**1.5)*(cfg['delta_rel'])**(((epoch-all_class_max_labeled_unconf_epoch[_class_])/cfg['epochs'])**1)

We are changing here per epoch than iterartion

**Comment by haruishi43** [comment]
So, since `total_all_class_labeled_unconf` isn’t updated, will `all_class_max_labeled_unconf_epoch` be all zeros?

Also, for `all_class_delta_rel`, it’ll be simplified to `+= (((max_val - last_val) / max_val)**1.5)*(cfg[‘delta_rel’]) ** ((epoch / cfg[‘epochs’])) right?

By the way, I’m curious about how you managed to achieve that 78.5 score with ResNet101 on Cityscapes 1/16 split. Could you share your exact hyperparameters and the code you used (e.g. loss weight annealing) to get that result like you said in #7 ?

**Comment by prantikhowlader** [comment]
Thanks, I will have a look at the hyperparameters used. I am inundated with deadlines in the next couple of months. I will have a look at the specific parameters used and specify in the github

**Comment by haruishi43** [comment]
Thanks I'll wait until then.

By the way, it seems like `all_class_delta_rel` is not updated. The values doesn't seem to change. It might be because `max_val` is constantly lower than `avg_val`. It's all due to `total_all_class_labeled_unconf` not being updated for each iteration (it's just initialized at each epoch, but not updated). Did you delete the snippet where you update `total_all_class_labeled_unconf`?

**Comment by prantikhowlader** [comment]
I think so, I need to look into this after a few months. As now the field has moved to LLMs. I am not  working in this field anymore.  I will have a look at it once I can get over with this list of back to back deadlines

**Comment by haruishi43** [comment]
Thanks. Looking at the code will be helpful because currently, it doesn’t seem to work as expected, and we can’t reproduce the results.

**Comment by prantikhowlader** [comment]
For running on for 1/16 split Resnet 101. the hyperparameters are:

Beta= 0.25(0-50 epochs), 0.10(50-100 epochs), after which 0.003
learning rate= 0.005
momentum=0.75
weight_decay=1e-3
scheduler=ReduceLROnPlateau[factor=0.1, threshold=1e-3]
